src/OpenTelemetry/Trace/Export/BatchingActivityProcessor.cs

```diff
@@ -0,0 +1,244 @@
+            {
+                this.ShutdownAsync(CancellationToken.None).ContinueWith(_ => { }).GetAwaiter().GetResult();
+            }
+
+            if (isDisposing)
+            {
+                if (this.exporter is IDisposable disposableExporter)
+                {
+                    try
+                    {
+                        disposableExporter.Dispose();
+                    }
+                    catch (Exception e)
+                    {
+                        OpenTelemetrySdkEventSource.Log.SpanProcessorException("Dispose", e);
+                    }
+                }
+            }
+        }
+
+        private async Task ExportBatchAsync(CancellationToken cancellationToken)
+        {
+            try
+            {
+                if (cancellationToken.IsCancellationRequested)
+                {
+                    return;
+                }
+
+                List<Activity> batch = null;
+                if (this.exportQueue.TryDequeue(out var nextActivity))
+                {
+                    Interlocked.Decrement(ref this.currentQueueSize);
+                    batch = new List<Activity> { nextActivity };
+                }
+                else
+                {
+                    // nothing in queue
+                    return;
+                }
+
+                while (batch.Count < this.maxExportBatchSize && this.exportQueue.TryDequeue(out nextActivity))
+                {
+                    Interlocked.Decrement(ref this.currentQueueSize);
+                    batch.Add(nextActivity);
+                }
+
+                var result = await this.exporter.ExportAsync(batch, cancellationToken).ConfigureAwait(false);
+                if (result != ExportResult.Success)
+                {
+                    OpenTelemetrySdkEventSource.Log.ExporterErrorResult(result);
+
+                    // we do not support retries for now and leave it up to exporter
+                    // as only exporter implementation knows how to retry: which items failed
+                    // and what is the reasonable policy for that exporter.
+                }
+            }
+            catch (Exception ex)
+            {
+                OpenTelemetrySdkEventSource.Log.SpanProcessorException(nameof(this.ExportBatchAsync), ex);
+            }
+        }
+
+        private async Task Worker(CancellationToken cancellationToken)
+        {
+            while (!cancellationToken.IsCancellationRequested)
+            {
+                var sw = Stopwatch.StartNew();
+                await this.ExportBatchAsync(cancellationToken).ConfigureAwait(false);
+
+                if (cancellationToken.IsCancellationRequested)
+                {
+                    return;
+                }
+
+                var remainingWait = this.scheduleDelay - sw.Elapsed;
+                if (remainingWait > TimeSpan.Zero)
+                {
+                    await Task.Delay(remainingWait, cancellationToken).ConfigureAwait(false);
```

Comment:
This is porting the span version so no need to change it here but it seems to me that exporting the data should be triggered not only on schedule but also on batch size.

Reply:
Agree, flushing the queue when the max queue size is reached is a good option. I've raised #704 to track it.